File: tools/src/symlinker.ts
// SPDX-License-Identifier: MPL-2.0

import * as path from "@std/path";
import { PATHS } from "./defines.ts";
import { Logger, createSymlink } from "./utils.ts";

const logger = new Logger("symlinker");

export function run(): void {
  const pairs: Array<[string, string]> = [
    [
      path.join(PATHS.loader_features, "link-features-chrome"),
      PATHS.features_chrome,
    ],
    [path.join(PATHS.loader_features, "link-i18n"), PATHS.i18n],
    [path.join(PATHS.loader_modules, "link-modules"), PATHS.modules],
  ];

  for (const [link, target] of pairs) {
    createSymlink(link, target);
  }

  logger.success("Symlinks created successfully.");
}

File: tools/lib/symlinker.rb
# SPDX-License-Identifier: MPL-2.0

require_relative "defines"
require_relative "utils"

module FelesBuild
  # loader の隣に、i18n と modules への近道を張る(build.rb symlink)
  module Symlinker
    LOGGER = Utils::Logger.new("symlinker")

    LINKS = {
      File.join(Defines::PATHS[:loader_modules], "link-modules") => Defines::PATHS[:modules],
    }.freeze

    # 張れなかったら、そこで止める。このまま進むと tsdown が
    # 「cannot find entry link-modules/**/*.mts」と言って転ぶ ── 近道が無いのが
    # 本当の理由なのに、そうは読めない。
    def self.run
      failed = LINKS.reject { |link, target| Utils.create_symlink(link, target) }
      unless failed.empty?
        raise "近道を張れなかった: #{failed.keys.join(', ')}\n" \
              "  この先で tsdown が entry を見つけられずに転ぶので、ここで止める。"
      end
      LOGGER.success "Symlinks created successfully."
    end
  end
end
